plutus-core/untyped-plutus-core/src/UntypedPlutusCore/Transform/Inline.hs

```diff
@@ -356,10 +356,17 @@ isPure = go True
             -- This case is not needed in PIR's `isPure`, because PIR's beta-reduction pass
             -- turns terms like this into `Let` bindings.
             Apply _ (LamAbs _ _ body) arg -> go True arg && go delayAndVarIsPure body
+            -- Applications can do work
```

Comment:
Maybe we could add an `Apply _ (Builtin ...` case?

Reply:
That's explicitly not handled, there's a note on the `Builtin` case that points to the note about differences to the PIR inliner.